### PR TITLE
Droth 2382 lane numbers on wrong track

### PR DESCRIPTION
**Pull request by ricardofmsimoes**
Fix the lane number representation in the frontend

### CODE REVIEW

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala [hunk 1]
```diff
@@ -195,20 +226,26 @@ class LaneFiller {
...
+
+          case (_, false) => Seq( createPersistedLane(MainLane.againstDirection, mainLane21SideCode.value, baseLane.municipalityCode, baseProps) )
 
-          case (true, false) => Seq( createPersistedLane(MainLane.againstDirection, SideCode.AgainstDigitizing.value, baseLane.municipalityCode, baseProps ) )
-          case (false, false) => Seq(
-                                createPersistedLane(MainLane.towardsDirection, SideCode.TowardsDigitizing.value, baseLane.municipalityCode, baseProps ),
-                                createPersistedLane(MainLane.againstDirection, SideCode.AgainstDigitizing.value, baseLane.municipalityCode, baseProps )
-                              )
+          case (false, false) => Seq( createPersistedLane(MainLane.towardsDirection,  mainLane11SideCode.value, baseLane.municipalityCode, baseProps),
+                                      createPersistedLane(MainLane.againstDirection, mainLane21SideCode.value, baseLane.municipalityCode, baseProps) )
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala]: Fixed.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala [hunk 2]
```diff
@@ -72,6 +73,27 @@ object LaneUtils {
...
+                                                        else SideCode.TowardsDigitizing
+
+                    case SideCode.TowardsDigitizing => if (laneCode.startsWith("1")) SideCode.TowardsDigitizing
+                                                        else SideCode.AgainstDigitizing
+
+                    case _ => SideCode.BothDirections
+                    }
+    }
+
+  }
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:94]: Move this to LaneService please.

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala [hunk 3]
```diff
@@ -171,8 +173,27 @@ class LaneFiller {
...
+                                                                  else SideCode.TowardsDigitizing
+
+                              case SideCode.TowardsDigitizing => if (laneCode.startsWith("1")) SideCode.TowardsDigitizing
+                                                                  else SideCode.AgainstDigitizing
+
+                              case _ => SideCode.BothDirections
+                            }
+      }
+
+  }
```
Comment (marquesfm) [inline digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala:194]: Please check the current method "getSideCode" existent at "TierekisteriImporterOperations" file.

We already perform something very similar to this when we import multiple assets from Tierekisteri using that information (RoadAddress Information).

Saying this I think we need to modify this code to something similar to:

```Scala
  def fixSideCode(roadAddress: BasicRoadAddress, laneCode: String): SideCode = {
    Track.apply(roadAddress.track) match {
      case Track.RightSide =>
        roadAddress.sideCode
      case Track.LeftSide =>
        roadAddress.sideCode match {
          case SideCode.TowardsDigitizing => SideCode.AgainstDigitizing
          case SideCode.AgainstDigitizing => SideCode.TowardsDigitizing
          case _ => SideCode.BothDirections
        }
      case _ =>
        SideCode.BothDirections
    }
  }

Comment (ricardofmsimoes) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala:194]: The logic inside TierekisteriImporterOperations is different.
For lanes we need a different logic.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala (hunk 2 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:94]: LaneFiller can't reach LaneUtils neither LaneService

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala (hunk 3 above)
Comment (marquesfm) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala:194]: Ok, with that information seems fine.

File: digiroad2-geo/src/test/scala/fi/liikennevirasto/digiroad2/lane/LaneFillerSpec.scala [hunk 4]
```diff
@@ -92,7 +93,9 @@ class LaneFillerSpec extends FunSuite with Matchers {
       roadLinkTowards2.linkId -> Seq( lane11 )
     )
 
-    val (filledTopology, changeSet) = laneFiller.fillTopology(topology, linearAssets)
+    val roadAddresses = Seq( BasicRoadAddress(roadLinkTowards2.linkId, 0, SideCode.TowardsDigitizing) )
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-geo/src/test/scala/fi/liikennevirasto/digiroad2/lane/LaneFillerSpec.scala:96]: Done.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala [hunk 5]
```diff
@@ -1,21 +1,22 @@
...
 import fi.liikennevirasto.digiroad2.service.lane.LaneService
 import fi.liikennevirasto.digiroad2.service.{RoadAddressService, RoadLinkService}
 import fi.liikennevirasto.digiroad2.{DummyEventBus, DummySerializer}
 import org.apache.http.impl.client.HttpClientBuilder
 import org.joda.time.DateTime
 
+
 case class RoadAddressesAux( linkId: Long, roadNumber: Long, roadPart: Long, track: Track, municipalityCode: Int,
-                          startMValue: Double, endMValue: Double, startAddrMValue:Long, endAddrMValue: Long)
+                             sideCode: SideCode, startMValue: Double, endMValue: Double, startAddrMValue:Long, endAddrMValue: Long)
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:19]: Done.

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala (hunk 2 above)
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:94]: Moved as requested

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala [hunk 6]
```diff
@@ -84,15 +106,16 @@ object LaneUtils {
...
                                             )
       // Get the road address information from our DB and convert the data to RoadAddressesAux
       val vkmRoadAddress = roadLinkTempDAO.getByRoadNumberRoadPartTrack(laneRoadAddressInfo.roadNumber.toInt, laneRoadAddressInfo.track, roadParts.toSet)
-                                          .map(elem => RoadAddressesAux(elem.linkId, elem.road, elem.roadPart,
-                                            elem.track, elem.municipalityCode.getOrElse(0), elem.startMValue, elem.endMValue,
-                                            elem.startAddressM, elem.endAddressM)
+                                          .map(elem => RoadAddressesAux(elem.linkId, elem.road, elem.roadPart, elem.track,
+                                            elem.municipalityCode.getOrElse(0), elem.sideCode.getOrElse(SideCode.BothDirections),
+                                            elem.startMValue, elem.endMValue, elem.startAddressM, elem.endAddressM)
                                           )
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:119]: Applied as requested

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala [hunk 7]
```diff
@@ -208,13 +231,15 @@ object LaneUtils {
           val endDifferenceAddr = road.endAddrMValue - laneRoadAddressInfo.endDistance
           val endPoint = if (isMainLane || endDifferenceAddr <= 0) road.endMValue else road.endMValue - endDifferenceAddr
 
-          calculateStartAndEndPoint(road, startPoint, endPoint) match{
-            case (start: Double, end: Double) =>  Some(PersistedLane(0, road.linkId, sideCode, laneCode, road.municipalityCode,
-                                                      start, end, Some(username), Some(DateTime.now()),
-                                                      None, None, expired = false,
-                                                      vvhTimeStamp, None, lane.properties))
+          val finalSideCode = fixSideCode( road, laneCode.toString )
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala:234]: Applied as requested

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 8]
```diff
@@ -101,8 +101,21 @@ trait LaneOperations {
...
+                                              elem.startAddrMValue, elem.endAddrMValue, elem.startMValue, elem.endMValue,elem.geom, Some(elem.sideCode), Some(0) )
+                                            )
+
+    val allLanes = assetsOnChangedLinks.filterNot(a => projectedLanes.exists(_.linkId == a.linkId)) ++ projectedLanes ++ lanesWithoutChangedLinks
+    val roadLinksToAdjust = roadLinks.filter( road =>  allLanes.exists(_.linkId == road.linkId))
+
+    val (lanesAfterSideCodeAdjust, changeSetInterm) = roadLinksToAdjust.foldLeft(Seq.empty[PersistedLane], changedSet) {
+      case ((_,changedSet), roadLink) =>
+        adjustLanesSideCodes(roadLink, allLanes, changedSet, roadAddresses)
+    }
```
Comment (marquesfm) [inline digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala:115]: Put this inside a method to try to maintain getLanesByRoadLinks as clean as possible.

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala [hunk 9]
```diff
@@ -171,83 +172,27 @@ class LaneFiller {
...
-                    else
-                      Seq()
-
-        val toRemove = lanesToProcess.filter( lane => (lane.laneCode >= MainLane.towardsDirection && lane.laneCode <= FourthRightAdditional.towardsDirection) ||
-                                              lane.sideCode != SideCode.BothDirections.value)
-                                      .map(_.id)
-                                      .filterNot(_ == 0L)
-
-        val lanesToAdd = (lanes ++ toAdd).filterNot(lane => toRemove.contains(lane.id))
+                              case _ => SideCode.BothDirections
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala:189]: Sicode can have the value BothDirection...

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala [hunk 10]
```diff
@@ -23,23 +23,24 @@ object  LaneFiller {
 
 
   case class SegmentPiece(laneId: Long, startM: Double, endM: Double, sideCode: SideCode, value: Seq[LaneProperty])
+
+  case class BasicRoadAddress(linkId: Long, track: Int, sideCode: SideCode )
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala:27]: Forget to remove!
Code not in use!

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 11]
```diff
@@ -131,13 +152,134 @@ trait LaneOperations {
...
+    if (lanes.isEmpty)
+      return (lanes, changeSet)
+
+    val filteredRoadAddresses = roadAddresses.find(_.linkId == roadLink.linkId)
+
+    val (mainLane11SideCode, mainLane21SideCode) = if (filteredRoadAddresses.nonEmpty)
+                                                    (fixSideCode(filteredRoadAddresses.head, MainLane.towardsDirection.toString),
+                                                      fixSideCode(filteredRoadAddresses.head, MainLane.againstDirection.toString))
+                                                  else
+                                                    (SideCode.TowardsDigitizing, SideCode.AgainstDigitizing)
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala]: Done

File: digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala [hunk 12]
```diff
@@ -131,13 +152,134 @@ trait LaneOperations {
...
+
+    val (mainLaneDirectionOK, mainLaneDirectionRemove) = getLaneDirectionOkAndToRemove ( filteredRoadAddresses.toSeq)
```
[the remark this answers is not in the document]
Comment (ricardofmsimoes) [reply digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala]: Done

File: digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (marquesfm) [reply digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/lane/LaneFiller.scala]: Already deleted, this conversation will be resolved.

**Pull request merged**